### ISSUE TITLE
[Neon.Common]: AsyncMutex is allowing multiple tasks to run

### ISSUE THREAD

**Issue by jefflill** [nforgeio/neonKUBE#550]
I'm seeing this exception when running the Cadence `Test_Emulation,Worker` test:

```
Message: System.InvalidOperationException : A task may only be disposed if it is in a completion state (RanToCompletion, Faulted or Canceled).
```

After setting some breakpoints I'm seeing the code protected by the async lock in `CadenceClient.StartWorkflowWorkerAsync()` being executed in parallel by two separate tasks.  This should never happen.

Investigate a fix or perhaps swap out our custom async implementations for official ones (if any).

**Comment by jefflill** [comment]
Microsoft has a Visual Studio related repo with some useful looking async primitives (with the MIT license):

https://github.com/microsoft/vs-threading

This also looks interesting because it supports recursion using a lock without deadlocking:

https://neosmart.net/blog/2017/asynclock-an-asyncawait-friendly-locking-library-for-c-and-net/

**Comment by jefflill** [comment]
After investigating, this is not a problem with the `AsyncMutex` implementation (which would have been shocking after all this time).  The problem was in Cadence code that wasn't actually awaiting the `AsyncMutex.AcquireAsync()` method.  The code was doing this:

```
using (mutex.AcquireAsync())
{
    ...
}
```

instead of this:

```
using (await mutex.AcquireAsync())
{
    ...
}
```

This mistake is super easy to make and very hard to see during code reviews and debugging.  The problem is that `Task` and the `AsyncMutex.AcquireAsync()` result both return `IDisposable` which the `using` statements will happily consume.  I found this article discussing a nice solution:

https://neosmart.net/blog/2017/asynclock-an-asyncawait-friendly-locking-library-for-c-and-net/

The basic idea is to have `AsyncMutex.AcquireAsync()` return an awaitable that doesn't implement `IDisposable` so that omitting the `await` will result in a compiler error,  I did this by creating and using the new `NonDisposableTask` type.
